### PR TITLE
feat: mine new overlay

### PR DESCRIPTION
**Pull request by nugaon**
Allowing the overlay address mining even when nonce exists. That case, `localstore` and `kademlia-metrics` folders will be removed and `statestore` will be cleared to start a clean state in the new neighborhood. Before the wipe, 10 seconds wait time allows the user to change their mind.

### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues.

### CODE REVIEW

File: pkg/node/node.go [hunk 1]
```diff
@@ -1359,3 +1374,28 @@ func isChainEnabled(o *Options, swapEndpoint string, logger log.Logger) bool {
 	logger.Info("starting with an enabled chain backend")
 	return true // all other modes operate require chain enabled
 }
+
+// removeContent removes all files in path. Copied function from cmd/db.go
+func removeContent(path string) error {
```
Comment (istae) [inline pkg/node/node.go:1379]: best to move this func to pkg/util/ioutil and import it from there

File: pkg/node/node.go [hunk 2]
```diff
@@ -278,11 +279,25 @@ func NewBee(
 		return nil, fmt.Errorf("compute overlay address: %w", err)
 	}
 
-	if nonceExists && o.TargetNeighborhood != "" {
-		logger.Warning("an overlay has already been created before, skipping targeting the selected neighborhood")
-	}
+	if o.TargetNeighborhood != "" {
+		if nonceExists {
```
Comment (istae) [inline pkg/node/node.go:283]: if the target neighborhood is specified, on each restart, wouldn't the directories would be nuked?

File: pkg/node/node.go [hunk 3]
```diff
@@ -278,11 +279,25 @@ func NewBee(
...
+			for _, dir := range dirsToNuke {
+				err = removeContent(filepath.Join(o.DataDir, dir))
+				if err != nil {
+					return nil, fmt.Errorf("delete %s: %w", dir, err)
+				}
+			}
+		}
 		// mine the overlay
 		targetNeighborhood := o.TargetNeighborhood
 		if o.TargetNeighborhood == "" && o.NeighborhoodSuggester != "" {
```
Comment (istae) [inline pkg/node/node.go:303]: for the suggester to run. the target neighborhood has to be "",  but this check is inside the `if o.TargetNeighborhood != "" {` check on line 283, no?

**Review by istae — CHANGES_REQUESTED** [review]
in order to achieve overlay change during runtime, i think you have to save the previous target-neighborhood in the statestore and only nuke the dirs (and change the overlay) if the target-neighborhood of the config option is different from the statestore one.

**Comment by ldeffenb** [comment]
> save the previous target-neighborhood in the statestore

Couldn't you just check the saved overlay (if there is one) to see if it is already in the specified target neighborhood?

File: pkg/node/node.go (hunk 3 above)
Comment (nugaon) [reply pkg/node/node.go:303]: correct, good catch! : ) move outside of the block

**Comment by nugaon** [comment]
Thanks for the comments regarding the missed use-case when user reloads the client with the mining param.

I corrected that and it also sets `TargetNeighborhood` with `NeighborhoodSuggester` only if nonce was not defined before.

**Review by istae — COMMENTED** [review]
some comments around the code changes would help
also let's post screenshots of the different test cases

File: pkg/node/node.go [hunk 4]
```diff
@@ -278,37 +278,58 @@ func NewBee(
...
+				time.Sleep(10 * time.Second)
+
+				const (
+					localstore = "localstore"
+					kademlia   = "kademlia-metrics"
+					statestore = "statestore"
+				)
+				dirsToNuke := []string{localstore, kademlia, statestore}
+				for _, dir := range dirsToNuke {
+					err = ioutil.RemoveContent(filepath.Join(o.DataDir, dir))
```
Comment (janos) [inline pkg/node/node.go:309]: It would be better to scope this `err` value in the loop instead to the outer variable which is currently two ifs above.

File: pkg/node/node.go [hunk 5]
```diff
@@ -278,37 +278,58 @@ func NewBee(
...
-			}
+	if targetNeighborhood != "" {
+		neighborhood, err := swarm.ParseBitStrAddress(targetNeighborhood)
+		if err != nil {
+			return nil, fmt.Errorf("invalid neighborhood. %s", targetNeighborhood)
 		}
 
-		if targetNeighborhood != "" {
+		if swarm.Proximity(swarmAddress.Bytes(), neighborhood.Bytes()) < uint8(len(targetNeighborhood)) {
+			if nonceExists {
```
Comment (janos) [inline pkg/node/node.go:297]: Removing directory before the `nbhdutil.MineOverlay`, which can fail, may result with only removed data.

File: pkg/node/node.go [hunk 6]
```diff
@@ -278,37 +278,58 @@ func NewBee(
...
+				logger.Info("Override nonce %d and clean state for neighborhood %s", nonce, targetNeighborhood)
+				logger.Warning("you have another 10 seconds to change your mind and kill this process with CTRL-C...")
+				time.Sleep(10 * time.Second)
+
+				const (
+					localstore = "localstore"
+					kademlia   = "kademlia-metrics"
+					statestore = "statestore"
```
Comment (janos) [inline pkg/node/node.go:305]: TLDR The cleanest way is instead to remove directories, it would be better for these services to implement its own internal nuke methods (that handle file descriptors and internal storage reinitalizations) and call them here, instead to remove their files.

Statestore directory is removed, but a few lines bellow statestore is passed to the setOverlay function. I would assume that the statestore is no longer reliable after its files are removed. Special attention should be made to ensure when interfering with storage internals outside of the services that manage their state. In this case, statestore must be reinitialised (and passed again to all already initialized services that use statestore) or if possible initilizad after the data on disk is removed. Handling nuke on node startup is different than in a separate nuke cli command as the node continues to run.

Nuking buy removing data directories is a questionable operation as these paths are constructed locally by assumption that they will not change in other places of the code where services that store data in them are initialized. I know that this is a mistake taken from the nuke cli command, but it would be good to make new code better.

Comment (nugaon) [reply pkg/node/node.go:302]: good point, I reinited statestore after the wipe as it is needed to fetch `nonce` before wipe.

Regarding the nuking, no service uses these databases before this point, all affected DB initializations happen in this function. It should have a separate `clearDb` method with options for sure if some of the data was required to be kept (please let me know if there is any). 
so basically, this is the only place where databases should managed as files, shouldn't it?

Comment (janos) [reply pkg/node/node.go:302]: Thank you for fixing it.

The statestore as a general key/value store, is a tricky to clean. I believe that this is a good time to plan something like a consolidation of all keys (or key prefixes) that are stored. It would be good to remove only the required keys.

I would say that it would even better to have a Nuke function on the databases to cleanup internal storage, in this case leveldb and not to worry about the file management. This would require a few more changes, but I believe that would result in more reliable code, not to worry about the files, just databases as abstraction. This can be an improvement later.

Comment (istae) [reply pkg/node/node.go:302]: the statestore actually has a nuke function but it preserves keys related to maintaing the overlay addresses, it may be improved to remove those entries when called upon.

Comment (janos) [reply pkg/node/node.go:302]: It is much better to adapt StateStoreAdapter.Nuke for cleaning. I think that it would be the safer way, and maybe to have the list of preserved prefixes a bit more transparent.

File: pkg/node/node.go [hunk 7]
```diff
@@ -278,37 +268,71 @@ func NewBee(
...
+				)
+				dirsToNuke := []string{localstore, kademlia}
+				for _, dir := range dirsToNuke {
+					err := ioutil.RemoveContent(filepath.Join(o.DataDir, dir))
+					if err != nil {
+						return nil, fmt.Errorf("delete %s: %w", dir, err)
+					}
+				}
+
+				stateStore.ClearForHopping()
```
Comment (istae) [inline pkg/node/node.go:311]: error checking?

File: pkg/node/node.go [hunk 8]
```diff
@@ -278,37 +278,58 @@ func NewBee(
...
 
-		if targetNeighborhood != "" {
+		if swarm.Proximity(swarmAddress.Bytes(), neighborhood.Bytes()) < uint8(len(targetNeighborhood)) {
+			if nonceExists {
+				logger.Info("Override nonce %d and clean state for neighborhood %s", nonce, targetNeighborhood)
+				logger.Warning("you have another 10 seconds to change your mind and kill this process with CTRL-C...")
+				time.Sleep(10 * time.Second)
+
+				const (
+					localstore = "localstore"
```
[the remark this answers is not in the document]
Comment (nugaon) [reply pkg/node/node.go]: yes, reasonable to make constants for that. I moved localstorage and kademlia path constants to ioutil within utils package.

File: pkg/util/ioutil/ioutil.go [hunk 9]
```diff
@@ -4,6 +4,18 @@
...
 
+import (
+	"errors"
+	"os"
+	"path/filepath"
+)
+
+// DB folders paths from bee datadir
+const (
+	DataPathLocalstore = "localstore"
```
[the remark this answers is not in the document]
Comment (nugaon) [reply pkg/util/ioutil/ioutil.go:15]: maybe `swarm.go`?

**Review by janos — APPROVED** [review]
LGTM. Thank you for addressing the comments @nugaon.

**Review by istae — APPROVED** [review]

**Review by martinconic — APPROVED** [review]

**Review by acha-bill — APPROVED** [review]
